api/rest/restapi.go

```diff
@@ -585,6 +585,62 @@ func (api *API) allocationHandler(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+// defined groups / aliases for filtering
+var filterGroups = map[string][]string{
+	"error":  {"error", "pin_error", "unpin_error", "cluster_error"},
+	"queued": {"queued", "pin_queued", "unpin_queued"},
+}
+
+func parseAllFilters(rawFilters string) map[string]bool {
+	// given the initial string of filters, return a slice of all individual strings
+	// to match on. This includes replacing alises
+	rawFilterList := strings.Split(strings.ToLower(rawFilters), ",")
+	var filterList = map[string]bool{}
+
+	for _, rawFilter := range rawFilterList {
+		if aliasSlice, exists := filterGroups[rawFilter]; exists {
+			for _, alias := range aliasSlice {
+				filterList[alias] = true
+			}
+		} else {
+			filterList[rawFilter] = true
+		}
+	}
+	return filterList
+
+}
+
+func filterStatusAll(filter string, pinInfos []types.GlobalPinInfoSerial) []types.GlobalPinInfoSerial {
+	if filter == "" {
+		return pinInfos
+	}
+	filterList := parseAllFilters(filter)
+	var filtered []types.GlobalPinInfoSerial
+NEXTPI:
+	for _, pi := range pinInfos {
+		for _, entry := range pi.PeerMap {
+			if _, exists := filterList[entry.Status]; exists {
+				filtered = append(filtered, pi)
+				continue NEXTPI
+			}
+		}
+	}
+	return filtered
+}
+
+func filterStatus(filter string, pinInfos []types.PinInfoSerial) []types.PinInfoSerial {
+	if filter == "" {
+		return pinInfos
+	}
+	filterList := parseAllFilters(filter)
+	for i := len(pinInfos) - 1; i >= 0; i-- {
+		if _, exists := filterList[pinInfos[i].Status]; !exists {
+			pinInfos = append(pinInfos[:i], pinInfos[i+1:]...)
```

Comment:
I was adding another test for just the `filterStatus` function and it kept failing even though `TestAPIStatusAllFilter` was passing. It finally clicked that my input slice, which was being passed to the two test cases, was being directly edited. So the first test case was passing, `error alias`, but the input slice then only contained the results of the `error alias` and so the second use case then failed as there were no `pinnin` statuses present. 

Generally in Go, if you want to edit a slice/array directly, that is fine but you just don't return a result of the same type, for example [`json.Unmarshal`](https://golang.org/pkg/encoding/json/#example_Unmarshal). But if you want to return the filtered result, you shouldn't edit the input slice, either by making a copy or the way it is done in `filterStatusAll`. See this [post](https://blog.golang.org/go-slices-usage-and-internals) for more information about how slices work.

Reply:
@lanzafame While I was applying this fix I ran into a little confusion about the way in which the filter should work with a CID specified (specifically in statusHandler() as opposed to statusAllHandler() ). In the previous commits I excluded filtering in this case for two reasons: 1) I did not know how to properly return a null result from the handler function when there were no matches. (I could likely figure this out myself with some work) and 2) I am not sure this is the desired behavior for a filter. To elaborate on (2), with one CID specified, the user would specify a filter for example 'pin_queued'. They would see one of two results; either the single CID with all peers if even one peer matched, or nothing at all. This is consistent with the behavior of the filter for statusAllHandler() but does not seem to be very useful to the user? Let me know if I should proceed with this logic or only apply filter to statusAllHandler() for now.

Reply:
From the user's perspective, if calling `filterStatus` it should be under the assumption that the query is being run against the peer that they are connected to, i.e.
```sh
$ ssh user@cluster-peer-1
$ ipfs-cluster-ctl --filter 'pin_queued' status QmdGd4C72HGpeGPLxUtDJMSxmgu7PxKYUbEzkfCvyJxdRX
QmdGd4C72HGpeGPLxUtDJMSxmgu7PxKYUbEzkfCvyJxdRX :
    > Peer QmSCZtfARHfCG6Ai8CYvfpNCTNudGa9UkajFuJNVRyDZaq : PIN_QUEUED | 2018-08-10T01:43:15Z
$ ssh user@cluster-peer-9
$ ipfs-cluster-ctl --filter 'pin_queued' status QmdGd4C72HGpeGPLxUtDJMSxmgu7PxKYUbEzkfCvyJxdRX
<no output because it isn't being pinned on this peer>
```

Does that clarify the issue?